File: windows/WallpaperInfoApp/WallpaperInfoApp/PlatformTimer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WallpaperInfoApp {

	class PlatformTimer {

		public PlatformTimer(WallpaperService wallpaperService) {
			_wallpaperService = wallpaperService;
		}

		private void scheduleSwitchWallpaper(int seconds) {
			if (_switchHandler != null) {
				cancelSwitchWallpaper();
			}
			_switchHandler = new System.Timers.Timer(seconds*1000);
			_switchHandler.Elapsed += (sender, args) => {
				// Console.WriteLine("PlatformTimer.eventHandler{}");
				if (!_pause) {
					_wallpaperService.wallpaperSwitchCallback();
				}
			};
			_pause = false;
			_switchHandler.Start();
		}
    
		void cancelSwitchWallpaper() {
			if (_switchHandler != null) {
				_switchHandler.Stop();
				_switchHandler = null;
			}
			_pause = true;
		}
    
		public void resetTimer() {
			if (!_pause) {
				resume();
			}
		}
    
		public void pause() {
			// print("PlatformTimer.pause()")
			cancelSwitchWallpaper();
		}
    
		public void resume() {
			// print("PlatformTimer.resume()")
			scheduleSwitchWallpaper(seconds: _interval);
		}
    
		public void setInterval(int interval) {
			_interval = interval;
			if (!_pause) {
				resume();
			}
		}

		private WallpaperService _wallpaperService;
		private System.Timers.Timer _switchHandler = null;
		private int _interval = 15;
		private bool _pause = false;
	}
}

File: apps/desktop-ubuntu/WallpaperInfoApp/PlatformTimer.py
#!/usr/bin/env python3

from threading import Timer

class RepeatingTimer(Timer):
    def run(self):
        while not self.finished.wait(self.interval):
            self.function(*self.args, **self.kwargs)

class PlatformTimer:
    _switchHandler = None
    _pause = False
    _interval = 5

    def __init__(self, wallpaperService):
        self._wallpaperService = wallpaperService
        
    def scheduleSwitchWallpaper(self, seconds):
        if (self._switchHandler != None):
            self.cancelSwitchWallpaper()
        self._switchHandler = RepeatingTimer(interval=seconds, function=self.executeCallback)
        self._pause = False
        self._switchHandler.start()

    def executeCallback(self):
        if not self._pause:
            self._wallpaperService.wallpaperSwitchCallback()

    def cancelSwitchWallpaper(self):
        if (self._switchHandler != None):
            self._switchHandler.cancel()
            self._switchHandler = None
        self._pause = True

    def resetTimer(self):
        if not self._pause:
            self.resume()
    
    def pause(self):
        self.cancelSwitchWallpaper()

    def resume(self):
        self.scheduleSwitchWallpaper(self._interval)

    def setInterval(self, interval):
        self._interval = interval
        if not self._pause:
            self.resume()
